util/dvsim/Scheduler.py

```diff
@@ -233,44 +104,372 @@ def on_sigint(signal_received, frame):
+        '''
+
+        changed = False
+
+        for target in self._scheduled:
+            if target in self._completed:
+                continue
+
+            to_pass = []
+            to_fail = []
+            for item in self._running[target]:
+                status = item.launcher.poll()
+                assert status in ['D', 'P', 'F']
+                if status == 'D':
+                    # Still running
+                    continue
+                elif status == 'P':
+                    log.log(VERBOSE, "[%s]: [%s]: [status] [%s: P]", hms,
+                            target, item.full_name)
+                    to_pass.append(item)
+                else:
+                    log.error("[%s]: [%s]: [status] [%s: F]", hms, target,
+                              item.full_name)
+                    to_fail.append(item)
+
+            for item in to_pass:
+                self._passed[target].add(item)
+                self._running[target].remove(item)
+                self.item_to_status[item] = 'P'
+                self._enqueue_successors(item)
+
+            for item in to_fail:
+                self._failed[target].add(item)
+                self._running[target].remove(item)
+                self.item_to_status[item] = 'F'
+                # It may be possible that a failed item's successor may not
+                # need all of its dependents to pass (if it has other dependent
+                # jobs). Hence we enqueue all successors rather than canceling
+                # them right here. We leave it to `_dispatch()` to figure out
+                # whether an enqueued item can be run or not.
+                self._enqueue_successors(item)
+
+            changed = changed or to_pass or to_fail
+
+        return changed
+
+    def _dispatch(self, hms):
+        '''Dispatch some queued items if possible.'''
+
+        # Compute sum of lengths of all lists in the given dict.
+        def __sum(d):
+            return sum([len(d[k]) for k in d])
+
+        num_slots = min(Scheduler.max_parallel - __sum(self._running),
+                        __sum(self._queued))
+        if num_slots <= 0:
+            return
+
+        for target in self._scheduled:
+            if target in self._completed:
+                continue
+
+            num_slots_per_target = min(
+                num_slots, Scheduler.max_parallel - len(self._running[target]),
+                len(self._queued[target]))
+            if num_slots_per_target <= 0:
+                continue
+
+            to_dispatch = []
+            while num_slots_per_target > 0:
+                # If no items remaining, then move onto the next target.
+                try:
+                    next_item = self._queued[target].pop(0)
+                except IndexError:
+                    break
+
+                if not self._ok_to_run(next_item):
+                    self._cancel_item(next_item, cancel_successors=False)
+                    self._enqueue_successors(next_item)
```

Comment:
I think this logic should probably live in `_enqueue_successors`. That way, we can choose never to enqueue stuff that should be cancelled. This avoids us having to wait a polling loop per target if everything is broken.

Reply:
Oh, I see. Good point!